### ISSUE TITLE
Add osm.org authentication to Wiki

### ISSUE THREAD

**Issue by lectrician1** [openstreetmap/operations#507]
_[Talk:Wiki discussion](https://wiki.openstreetmap.org/wiki/Talk:Wiki#Authentication_from_osm.org)_

[PluggableAuth](https://www.mediawiki.org/wiki/Extension:PluggableAuth) looks like a suitable extension authenticating users using OAuth from osm.org to the Wiki.

You're probably going to have to read into the specifics on your own since the various options for how users can log in can be a bit confusing. 

Here is an example of a [Wiki that uses it](https://www.freeipa.org/index.php?title=Special:UserLogin&returnto=Main+Page).

**Comment by Firefishy** [comment]
The users on the osm.org and wiki are not 1:1 matched. Any proposal to resolve the discrepancy?

**Comment by tomhughes** [comment]
That is basically why the idea has been deemed a non-starter in the past.

**Comment by lectrician1** [comment]
The extension matches users that have the same username/email on osm.org and the wiki (not sure if both or just 1).

If users want to "link" their accounts, they're going to have to make sure these are the same.

Fortunately users can change their username and email on the wiki and osm.org, so they get to decide.

**Comment by tomhughes** [comment]
So how does it stop me taking over a wiki account by creating an OSM account with the same name?

**Comment by lectrician1** [comment]
That's why I think both have to be the same.

**Comment by tomhughes** [comment]
That was my point, if just have the same name is all it takes then I can create an account call JohnBrown on OSM and then login to the wiki as JohnBrown even if there is already an account called that which isn't mine?

**Comment by lectrician1** [comment]
Oh no, what I meant is that both the osm.org and wiki email **and** username have to be the same.

**Comment by mmd-osm** [comment]
I wonder if there’s an option for an already logged on wiki user to associate their account with an external identity provider, such as the osm.org website. Of course the user would also have to log on to osm.org to establish that link.

I think user names don’t have to match in that case, as long as a wiki user has an external alias identity. This might also include other trustworthy identity providers.

**Comment by maro-21** [comment]
What if I wanted to create a bot account on Wiki - will I have to create an OSM account for it first?

**Comment by lectrician1** [comment]
> What if I wanted to create a bot account on Wiki - will I have to create an OSM account for it first?

No. You can continue to create individual accounts on both osm.org and the Wiki.

**Comment by lectrician1** [comment]
> I wonder if there’s an option for an already logged on wiki user to associate their account with an external identity provider, such as the osm.org website.

That's what this extension does.

> I think user names don’t have to match in that case, as long as a wiki user has an external alias identity. This might also include other trustworthy identity providers.

What's an external alias identity? Just any account?

**Comment by tomhughes** [comment]
He's talking about a system like osm.org uses, where if you setup login with Google or Facebook or whatever we store the identity of that account so that you can login with it without your email and/or username having to match.

In other words the user has to explicitly associate the accounts and cause a link to be stored rather than an implicit link based on name and email.

**Comment by lectrician1** [comment]
What really happens is that a button below the normal "log in" button on the wiki will display with the label "Log in with OpenStreetMap.org"

The user can then click that and the external identification OAuth window (idk entirely how OAuth works) appears. 

Whatever the user logins as via that window, the OAuth client returns the username and email of that account to the wiki. If an account with that username and password exits on the Wiki, they are logged-in.

Users can still log in with their normal Wiki username and password in the current input boxes on the login page on the Wiki if they want to.

**Comment by mmd-osm** [comment]
So what you're describing here is essentially the same thing I've demoed for Discourse: https://user-images.githubusercontent.com/5842757/96035123-e1a1e700-0e62-11eb-9f05-fa92e8751ad0.gif
I've tried the same thing on a test wiki, and it wasn't all that smooth sailing.

Still this doesn't address the issue of associating osm and wiki accounts, when we can't rely on a matching name/email.

**Comment by tomhughes** [comment]
Well he is suggesting that we do rely on matching and that people can change their names on one or both systems to achieve that is necessary.

I would very much prefer an explicit link though I have to say.

**Comment by matkoniecz** [comment]
> The users on osm.org and wiki are not 1:1 matched. Any proposal to resolve the discrepancy?

(0) Establish username policy on wiki that you must have the same username for OSM account and Wiki account, handle mass renaming (requires agreement from people who would need to handle flood of renames - possibly find people who would get this rights temporarily, solely to handle this drudgery, I self nominate here)
(1) If OSM account and Wiki account has the same username and not the same email - contact users via wiki talk page and ask them to change email on wiki (is it OK from privacy standpoint? Maybe contact all users who have account name shared by the wiki username? I also self-nominate for help here)
(2) Ask again, say one month later
(3) Ask again, say one month later
(4) Ask again, say one month later, this time via PMing OSM usernames and warn that if it is their account then they should harmonize emails
(4) Forcefully rename wiki accounts with mismatching email and matching usernames (append "conflicting username" or something similar), I again self nominate here for drudgery (maybe rename using boit)
(5) Block wiki accounts with mismatching email and the same names that were renamed

Now users on osm.org and wiki are 1:1 matched.

Non exhaustive list of traps:

- what happens if someone changes displayed username on OSM website?
- needs agreement from OSM Wiki community
- needs people to handle drudgery of renaming and blocking
- some people will ignore all requests and will be confused and stop contributing
- in some cases it will result in person operating under identify used previously by someone else
- is renaming username freeing previous one?

Quite obnoxious, but I consider it to be worth doing.

**Comment by mmd-osm** [comment]
@matkoniecz : did you miss the part where we said an explicit link would be preferred?

**Comment by flohoff** [comment]
First question i would have is: Does mediawiki support all usernames we have in OSM. OSM has been very relaxed with user naming so there may be pitfalls with non-ascii, utf8-emoji, control character etc names.

**Comment by mmd-osm** [comment]
@flohoff : it doesn't matter, we won't be matching by user names anyway. That's the idea of having explicit links.

**Comment by Nautic** [comment]
**Wikimedia**: For your plan to synergize OSM.org and wiki.OSM.org  it would be great to consider Wikimedia-accounts too :-)
(Wikipedia, Commons, Wikidata, ... as global Wikimedia user account)
This would be a great synergy for both projects!

**Comment by Firefishy** [comment]
Added the "help-wanted" label. I can support any tests anyone wants to run.

**Comment by Firefishy** [comment]
Challenges we would likely need to solve to implement this ticket:

1.  osm.org "Display Names" (aka Usernames) are not fixed, we deal with this on community.osm.org by setting the name on community.osm.org on each user login and use the UserID from osm.org as the unique key. The oauth is specially blessed for community.osm.org to be able to access user email address.
2. Many usernames for wiki.osm.org do not match the same user on osm.org, how do we workaround / resolve this?
3. How would the process to migrate wiki users to osm.org based logins work? (Technically and practically)
4. Maintainability going forward, if custom code is used, it needs to be maintained going forward.

Maybe we could have a brainstorming video call sometime to draw out a plan?

**Comment by tomhughes** [comment]
You can if you want just allow the names to be different, as help.osm.org does and as osm.org does if you use Google or Facebook to login.

You just store the identity on the remote system as an attribute of the account, and when some proves control of that identity you let them login to that account.

**Comment by mmd-osm** [comment]
Last time I looked for some mediawiki plugins, I couldn't find anything similar to the approach we're using for Discourse.

I wonder if some development effort on the Rails port would be acceptable, if that's eliminating the need for a custom wiki plugin. We might check if https://www.mediawiki.org/wiki/Extension:OpenID_Connect could be an option for us.

**Comment by mmd-osm** [comment]
I played a bit with https://github.com/doorkeeper-gem/doorkeeper-openid_connect to enable Openid connect support on the Rails port. Most of the installation worked out of the box (as far as you can call this working). I'm documenting a few changes I did to the config. For sure this needs some more work, as it's meant as a very first step to get started. I believe this could be used a starting point fo enable OpenID Connect mediawiki extension. I haven't tried this out end-to-end yet, though.

### Setting up doorkeeper-openid_conect gem

After `rails generate doorkeeper:openid_connect:install`, I added  Rails.application.reloader.to_prepare do to the doorkeeper_openid_connect.rb file, otherwise the migration step would bail out.

SAFETY_ASSURED=1 was used to silence an otherwise unhappy strong_migration.

config/initializers/doorkeeper_openid_connect.rb:

<details>

```yml
# frozen_string_literal: true

Rails.application.reloader.to_prepare do

Doorkeeper::OpenidConnect.configure do
  issuer do |resource_owner, application|
     'http://localhost:3000'
  end

  signing_key <<~KEY
    -----BEGIN RSA PRIVATE KEY-----
    
[add private key...]

    -----END RSA PRIVATE KEY-----
  KEY

  subject_types_supported [:public]

  resource_owner_from_access_token do |access_token|
    # Example implementation:
    User.find_by(id: access_token.resource_owner_id)
  end

  auth_time_from_resource_owner do |resource_owner|
    # Example implementation:
    #resource_owner.current_sign_in_at
  end

  reauthenticate_resource_owner do |resource_owner, return_to|
    # Example implementation:
    # store_location_for resource_owner, return_to
    # sign_out resource_owner
    # redirect_to new_user_session_url
  end

  # Depending on your configuration, a DoubleRenderError could be raised
  # if render/redirect_to is called at some point before this callback is executed.
  # To avoid the DoubleRenderError, you could add these two lines at the beginning
  #  of this callback: (Reference: https://github.com/rails/rails/issues/25106)
  #   self.response_body = nil
  #   @_response_body = nil
  select_account_for_resource_owner do |resource_owner, return_to|
    # Example implementation:
    # store_location_for resource_owner, return_to
    # redirect_to account_select_url
  end

  subject do |resource_owner, application|
    # Example implementation:
    resource_owner.id

    # or if you need pairwise subject identifier, implement like below:
    # Digest::SHA256.hexdigest("#{resource_owner.id}#{URI.parse(application.redirect_uri).host}#{'your_secret_salt'}")
  end
 
  claims do
    claim :preferred_username, scope: :openid do |resource_owner, scopes, access_token|
      # Pass the resource_owner's preferred_username if the application has
      # `profile` scope access. Otherwise, provide a more generic alternative.
      resource_owner.display_name
    end

  end  
  
end

end
```

</details>

I added a new privileged scope 'openid' (translation for en.oauth.scopes.openid should also be added):

<details>

```rails
diff --git a/lib/oauth.rb b/lib/oauth.rb
index 7ff2ba8b4..1c7b38636 100644
--- a/lib/oauth.rb
+++ b/lib/oauth.rb
@@ -1,6 +1,6 @@
 module Oauth
   SCOPES = %w[read_prefs write_prefs write_diary write_api read_gpx write_gpx write_notes].freeze
-  PRIVILEGED_SCOPES = %w[read_email skip_authorization].freeze
+  PRIVILEGED_SCOPES = %w[read_email skip_authorization openid].freeze
 
   class Scope
     attr_reader :name
```

</details>

Skipping the suggested view generation (`rails generate doorkeeper:views`) and added the nonces field here instead:

<details>

```erb
diff --git a/app/views/oauth2_authorizations/new.html.erb b/app/views/oauth2_authorizations/new.html.erb
index 971e0e20a..ac9c7c6c5 100644
--- a/app/views/oauth2_authorizations/new.html.erb
+++ b/app/views/oauth2_authorizations/new.html.erb
@@ -18,6 +18,7 @@
       <%= f.hidden_field :state, :value => @pre_auth.state %>
       <%= f.hidden_field :response_type, :value => @pre_auth.response_type %>
       <%= f.hidden_field :scope, :value => @pre_auth.scope %>
+      <%= f.hidden_field :nonce, :value => @pre_auth.nonce %>
       <%= f.hidden_field :code_challenge, :value => @pre_auth.code_challenge %>
       <%= f.hidden_field :code_challenge_method, :value => @pre_auth.code_challenge_method %>
       <%= f.primary t(".authorize") %>
@@ -30,6 +31,7 @@
       <%= f.hidden_field :state, :value => @pre_auth.state %>
       <%= f.hidden_field :response_type, :value => @pre_auth.response_type %>
       <%= f.hidden_field :scope, :value => @pre_auth.scope %>
+      <%= f.hidden_field :nonce, :value => @pre_auth.nonce %>
       <%= f.hidden_field :code_challenge, :value => @pre_auth.code_challenge %>
       <%= f.hidden_field :code_challenge_method, :value => @pre_auth.code_challenge_method %>
       <%= f.submit t(".deny") %>
```

</details>

Adjusted routes to point to /oauth2 instead

<details>

```
diff --git a/config/routes.rb b/config/routes.rb
index e45d56701..b1e69dce3 100644
--- a/config/routes.rb
+++ b/config/routes.rb
@@ -1,4 +1,5 @@
 OpenStreetMap::Application.routes.draw do
+  use_doorkeeper_openid_connect :scope => "oauth2"
   use_doorkeeper :scope => "oauth2" do
     controllers :authorizations => "oauth2_authorizations",
                 :applications => "oauth2_applications",
```

</details>
                 
### Result

In the end, I used some demo OpenId client, and got the following result in the end when calling http://localhost:3000/oauth2/userinfo                 

```json
{
    "sub": "2",
    "preferred_username": "mmd3"
}
```

http://localhost:3000/.well-known/openid-configuration contents:

<details>

```json
{
  "issuer": "http://localhost:3000",
  "authorization_endpoint": "http://localhost:3000/oauth2/authorize",
  "token_endpoint": "http://localhost:3000/oauth2/token",
  "revocation_endpoint": "http://localhost:3000/oauth2/revoke",
  "introspection_endpoint": "http://localhost:3000/oauth2/introspect",
  "userinfo_endpoint": "http://localhost:3000/oauth2/userinfo",
  "jwks_uri": "http://localhost:3000/oauth2/discovery/keys",
  "scopes_supported": [
    "read_prefs",
    "write_prefs",
    "write_diary",
    "write_api",
    "read_gpx",
    "write_gpx",
    "write_notes",
    "read_email",
    "skip_authorization",
    "openid"
  ],
  "response_types_supported": [
    "code"
  ],
  "response_modes_supported": [
    "query",
    "fragment",
    "form_post"
  ],
  "grant_types_supported": [
    "authorization_code"
  ],
  "token_endpoint_auth_methods_supported": [
    "client_secret_basic",
    "client_secret_post"
  ],
  "subject_types_supported": [
    "public"
  ],
  "id_token_signing_alg_values_supported": [
    "RS256"
  ],
  "claim_types_supported": [
    "normal"
  ],
  "claims_supported": [
    "iss",
    "sub",
    "aud",
    "exp",
    "iat",
    "preferred_username"
  ]
}
```

</details>

**Comment by mmd-osm** [comment]
I put together a small video highlighting the end-to-end process. In theory there's a good chance that this might actually work...

<details>
<summary>Demo time (screencast)</summary>

mmd2 was an already existing wiki user with a matching email address. 


![wiki_logon](https://user-images.githubusercontent.com/5842757/183755460-9679d59b-f0bd-4033-9d9f-185983ca7eaa.gif)

</details>

**Comment by mmd-osm** [comment]
@not-my-profile : can you maybe write a few words on how you have planned to support your extension going forward? How about translating messages to other languages?

As mentioned on the community site, I had a strong focus to use an officially wiki extension and avoid any custom code (see my comment from Aug 5). Due to the extreme shortage of knowledgable MediaWiki folks around here, this seemed like the least risky option, although user experience may not be as good as it could be.

**Comment by not-my-profile** [comment]
I would maintain it in my [self-hosted repo](https://git.push-f.com/mw-redirect-auth/) and document it on the mediawiki.org wiki (I have not yet done so because I don't want to spend time writing documentation if I don't yet know if the extension will actually be used by the OSM wiki). I'd answer questions about the extension on its talk page at mediawiki.org as well as be available for contact via email for the operations working group. I think for translations I would sign up my extension at [translatewiki.net](https://translatewiki.net/) (which is used by MediaWiki itself as well as hundreds of extensions).

**Comment by mmd-osm** [comment]
Any chance @Firefishy, @tomhughes and @pnorman could look at this issue in one of the next OWG meetings and discuss whether the approach suggested by @not-my-profile would work at all for the OSM wiki?

I saw that you already started exploring evaluation criteria for wiki addons in #785. Maybe this could be a good opportunity to add a bit more detail there as well.

Why am I bringing this up now? A board member has recently mentioned in a conference talk that all these nice ideas on the _OSM Top Ten Task_ list are impossible to implement (OSM.org authentication to the Wiki being one of them). I think that's not quite accurate.

**Comment by tomhughes** [comment]
What exactly do you want OWG to discuss? I've only skimmed this thread but it seems to be OpenID Connect based which is not something osm.org support currently.

In any case it's not clear that there's anything here that needs a policy decision and that couldn't be actioned as a normal pull request proposing a change to the wiki if it's just a question of installing an extension?

**Comment by mmd-osm** [comment]
@not-my-profile has implemented a custom Mediawiki extension, that doesn't depend on OpenID Connect and is specifically tailored to the needs of OSM wiki user migration.

The alternative option I've mentioned earlier on uses stable Mediawiki extension, and depends on OIDC on openstreetmap-website, although with a less than optimal UX.

Policy wise my question would be, what is more important: use an off the shelf stable Wiki extension, and sacrifice user experience, or have something more custom, with a potential long term maintenance burden.

If that can be handled using the normal process, that is also fine.

**Comment by tomhughes** [comment]
OpenID Connect support is no w live on openstreetmap.org so we should be able to use https://www.mediawiki.org/wiki/Extension:OpenID_Connect to add Login with OSM to the wiki if we want.

To use OpenID Connect an OAuth 2 application needs to be registered with the new `openid` scope. Optionally the `read_email` scope can be enabled (for privileged applications owned by administrative users) to grant access to email addresses which we would probably want to do here.

**Comment by mmd-osm** [comment]
Can we maybe keep the test wiki https://test.wiki.openstreetmap.org/ around for a bit longer and give the OpenID connect extension a try? That’s of course only once the Unicode tests are done.

**Comment by 1ec5** [comment]
https://github.com/openstreetmap/operations/issues/507#issuecomment-1484224671:

> Policy wise my question would be, what is more important: use an off the shelf stable Wiki extension, and sacrifice user experience, or have something more custom, with a potential long term maintenance burden.

Where did we land on this question of using [OpenID Connect](https://www.mediawiki.org/wiki/Extension:OpenID_Connect) versus [RedirectAuth](https://git.push-f.com/mw-redirect-auth/about/)? [The forum discussion](https://community.openstreetmap.org/t/rfc-two-new-extensions-for-the-wiki-log-in-via-openstreetmap-org-and-vote-via-a-gui/4006) seems to have digressed into voting mechanisms. Voting may be the most acute pain point for the community, but the scope of this issue is a bit broader because we want mappers to [feel comfortable](https://community.openstreetmap.org/t/removal-of-from-to-tags-in-routes-if-the-name-already-contains-the-values/138323/46) contributing to the wiki as a whole too.

https://github.com/openstreetmap/operations/issues/507#issuecomment-1206368122:

> 1. osm.org "Display Names" (aka Usernames) are not fixed, we deal with this on community.osm.org by setting the name on community.osm.org on each user login and use the UserID from osm.org as the unique key. The oauth is specially blessed for community.osm.org to be able to access user email address.
> 2. Many usernames for wiki.osm.org do not match the same user on osm.org, how do we workaround / resolve this?

Do we actually need the wiki’s user name to match the OSM user’s display name? OpenHistoricalMap has been keying off the UID rather than the display name, and we’re letting the user choose a different user name anyways. Omitting the `read_email` scope (OpenHistoricalMap/issues#1076) means the user has to reenter their e-mail address when creating an account. I guess that would be the case for the wiki too, since we’ve configured Special:CreateAccount to require an e-mail address. But adding that scope would be nice.

If we just want the user names to match for our own sanity, wiki bureaucrats could set up a process for users to request renaming. If a user accidentally creates a redundant OpenID-authenticated account, a bureaucrat can easily make the connected account [usurp](https://www.mediawiki.org/wiki/Help:Renaming_users#Usurping_an_account) the unconnected account. If we really want to force the issue, we could do what Wikimedia did when introducing Single User Login: forcibly rename any unlinked accounts that conflict with linked accounts to e.g., `example~~commonswiki`; anyone who wants to reclaim the unlinked account afterwards can usurp one or the other.

> 3. How would the process to migrate wiki users to osm.org based logins work? (Technically and practically)

Do we even need to migrate existing wiki users? It doesn’t sound like RedirectAuth would require it. Similarly, PluggableAuth appears to support installing one or more authentication plugins while leaving local logins enabled. In order to prevent impersonation and provide accountability for voting, I think PluggableAuth lets us automatically assign OpenID-based accounts to a user group, which would serve as a public indication of how the user authenticates. I don’t know if it can prevent the same OSM UID from being linked to multiple sockpuppet wiki accounts, as RedirectAuth does.

https://github.com/openstreetmap/operations/issues/507#issuecomment-2953723852:

> Can we maybe keep the test wiki https://test.wiki.openstreetmap.org/ around for a bit longer and give the OpenID connect extension a try? That’s of course only once the Unicode tests are done.

This is a reference to #373. I agree that it would be good to get that out of the way first. OSM has some prominent users who use non-BMP characters in their display names. Automating wiki account creation based on those display names, even by accident, could lead to some awkwardness as the abuse filter I implemented in https://github.com/openstreetmap/operations/issues/583#issuecomment-983230684 prevents them from participating in discussions.

**Comment by tomhughes** [comment]
Well this is the first I've heard of RedirectAuth and looking at it I don't see how it's relevant - it looks like it's intended for use with standalone authentication servers that act as a sort of frontend for arbitrary sites but we don't have any technology on openstreetmap.org that would allow it to work? Though I'm kind of guessing at how it works given there is zero documentation!

**Comment by tomhughes** [comment]
I think it's clear by the way that we can't rely on name matching - there's no way we're going to do the kind of mass renaming that would need.

I'm surprised to learn from the forum discussion that the OIDC extension can only match on email as I assumed it would do most things do and link the OIDC subject (which is just the numeric OSM user ID value) to a wiki account.

**Comment by 1ec5** [comment]
> Well this is the first I've heard of RedirectAuth and looking at it I don't see how it's relevant - it looks like it's intended for use with standalone authentication servers that act as a sort of frontend for arbitrary sites but we don't have any technology on openstreetmap.org that would allow it to work? Though I'm kind of guessing at how it works given there is zero documentation!

RedirectAuth came up in https://github.com/openstreetmap/operations/issues/507#issuecomment-1278695638. (GitHub isn’t showing the comment by default because this issue has so many comments.) There used to be a live demo wiki back in 2022. It claims to use whatever osm-website had before OIDC, matching on the UID. The downside of course is that it’s a custom MediaWiki extension. We’re already currently stuck on some upgrades because of another custom extension whose author hasn’t been responsive.

> I'm surprised to learn from the forum discussion that the OIDC extension can only match on email as I assumed it would do most things do and link the OIDC subject (which is just the numeric OSM user ID value) to a wiki account.

I agree that the UID would be ideal. That’s what we set up for OHM due to the likelihood of users periodically changing their e-mail address on OSM: OpenHistoricalMap/ohm-website#293.

If I’m reading the documentation correctly, the OIDC extension only uses the e-mail address for migrating existing users, but it sounds like we don’t want to do that anyways. For linking accounts, I think the `providerConfig` and `preferredUsernameProcessor` options could be relevant. https://github.com/openstreetmap/operations/issues/507#issuecomment-1209868428 demonstrates passing the UID to MediaWiki, so it seems feasible somehow.

**Comment by tomhughes** [comment]
So if I'm reading that right it does use the subject to link the accounts, but when you first login using OpenID it uses the email to find the account to link to - after that you can presumably change either email and it will still work as it has the ID link then.

What's missing is a way to login to the wiki with your password and then link that to your OSM account without it having to match them on email the first time.

Certainly if it is only for the first linking that they need to match then that's less of an issue.

**Comment by 1ec5** [comment]
> What's missing is a way to login to the wiki with your password and then link that to your OSM account without it having to match them on email the first time.

This is what Special:LinkAccounts and Special:UnlinkAccounts are for. RedirectAuth apparently [supports it](https://community.openstreetmap.org/t/rfc-two-new-extensions-for-the-wiki-log-in-via-openstreetmap-org-and-vote-via-a-gui/4006/5). I don’t see any indication that PluggableAuth supports it, but I haven’t had a chance to try out either extension on a running MediaWiki instance.

**Comment by valerio-bozzolan** [comment]
What is needed here seems the modern auto-verification workflow, available since 2021 in the MediaWiki extension [WSOAuth](https://www.mediawiki.org/wiki/Extension:WSOAuth) (based on PluggableAuth).

So, the MediaWiki user can just do yet another manual login in the wiki, and can just click a button in their preferences, to manually ALSO login in the other identity provider, to connect the two accounts (whatever the nickname differences):

<img width="1034" height="651" alt="Image" src="https://github.com/user-attachments/assets/2bc40f66-1567-479f-9d2e-8a612c143f12" />

https://commons.wikimedia.org/wiki/File:MediaWiki_1.31_special_preferences_WSOAuth_extension_connect_remote_account_English.png

From that moment, next time, the MediaWiki user can enjoy the quick login with OAuth, so, not using the local MediaWiki password:

<img width="326" height="471" alt="Image" src="https://github.com/user-attachments/assets/96ef8983-2092-421d-89fa-855c16adbc05" />

https://commons.wikimedia.org/wiki/File:MediaWiki_1.31_special_login_WSOAuth_extension_English.png

P.S. if this workflow is not suitable for your needs, please leave your feedback in the related task on Wikimedia Phabricator. There you can find my specifications from 2021 and the kind code implementation from the kind developer Xxmarijnw, who was super-kind, professional and fast; so, I strongly suggest to contact again, in case any custom implementation is needed.

https://www.mediawiki.org/wiki/Extension:WSOAuth - extension

https://phabricator.wikimedia.org/T283908 - that feature I designed in 2021 and was funded by Wikimedia Switzerland and implemented by Xxmarijnw in the same year

https://www.mediawiki.org/wiki/User:Xxmarijnw - freelance developer, probably available for quotes

------

P.S. thanks to @Danysan1 (board Wikimedia Italy) from bringing this discussion to my attention. Following now since I'm very interested. Thanks

**Comment by mmd-osm** [comment]
I gave WSOAuth a try with a customOSM AuthProvider which is leveraging the API 0.6 /user/details.json endpoint.

The short video below highlights a use case where an existing Wikiuser is associated with an osm.org user. After logging off, the osm.org based logon can be used instead of entering the wiki user/password.

https://github.com/user-attachments/assets/e66552e0-8568-404a-9725-f63b1ea555df

User mapping is based on osm uid:

<img width="638" height="143" alt="Image" src="https://github.com/user-attachments/assets/a755a683-dbd6-4ea5-b34b-a2520e9e747b" />

If I try to "Login with OpenStreetMap" and there's an existing Wiki user with the same osm display_name, WSOAuth tries to append some number (e.g. "mmd" -> "mmd 1") until it finds an available user. In other words, existing Wiki users need to establish a link to osm.org from the preferences page, or they end up creating a different user.

**Comment by 1ec5** [comment]
> If I try to "Login with OpenStreetMap" and there's an existing Wiki user with the same osm display_name, WSOAuth tries to append some number (e.g. "mmd" -> "mmd 1") until it finds an available user. In other words, existing Wiki users need to establish a link to osm.org from the preferences page, or they end up creating a different user.

A wiki administrator can modify the messages on Special:UserLogin so that an existing wiki user will know to log in natively first. If they miss that bit and create a secondary account anyways, would that prevent them from subsequently associating their original wiki account with the same OSM account? In other words, is it possible for a single OSM account to be associated with multiple wiki accounts?

**Comment by valerio-bozzolan** [comment]
> In other words, is it possible for a single OSM account to be associated with multiple wiki accounts?

No. 1 account is connected to 1 account.

**Comment by mmd-osm** [comment]
Iirc, you can link multiple osm accounts to the same wiki account, but not the other way around. Maybe this is even a bug?

Re. secondary account creation:
I think it would be better if "login with osm" would check if the wiki user name hasn't been taken yet. Otherwise it should ask the user to login with username and password, then establish a new link to osm.org from preferences. 

Similar if the osm username is not    suitable for the wiki: let the user create the account manually, then establish a link through preferences.

Creating fallback user names with " 1", etc. suffix is most likely not desirable and only creates extra effort for sysadmins. Due to the lack of functionality to disassociate your wiki user from an osm.org user, there's no way a user could fix this on their own.

I'm wondering if we could/should include the email addresses for existing user matching as well.

**Comment by mmd-osm** [comment]
I did some further tweaks to my WSOAuth [fork](https://codeberg.org/mmd/WSOAuth) and introduced a new flag to control fallback username generation. If the Wiki username is already taken, the following message is shown instead of creating a secondary account:

Wording is still a bit preliminary. Creating a new account instead could also be mentioned there.

<img width="726" height="628" alt="Image" src="https://github.com/user-attachments/assets/86a3ded5-4b16-4e99-86d8-7e3102227075" />

**Comment by matkoniecz** [comment]
How cases where someone else created wiki account with username of active OSM user should be handled?

Forcefully rename them?

**Comment by mmd-osm** [comment]
Renaming accounts is up to bureaucrats to decide on a case by case basis. It will always be a manual process and not in scope of this extension.

Obviously users can still create another wiki account and link it to their differently named osm account.

**Comment by 1ec5** [comment]
The equivalent on Wikimedia wikis is called [usurpation](https://www.mediawiki.org/wiki/Help:Renaming_users#Usurping_an_account) and only happens by request.

I suspect we’d have more cases of users who have different user names between OSM and the wiki, who would then create a second wiki account through the new flow because they didn’t heed the warning. Matching e-mail addresses as a fallback might help, but even then, anything is possible.

There probably aren’t too many people who would feel a great loss about having to change their wiki identity as a result. For those who do, we could install the [UserMerge](https://www.mediawiki.org/wiki/Extension:UserMerge) extension.

**Comment by mmd-osm** [comment]
UserMerge is definitely an option. However, it requires some more research and testing. I doubt it will work out of the box with WSOAuth, leaving us with stale mapping entries in the `wsoauth_multiauth_mappings` table.[^1]

-> Edit: CONFIRMED: merging newly created osm.org-based user into an exsiting wiki  user acount does not work with UserMerge. Requires manual intervention on db, e.g.: `update wsoauth_multiauth_mappings set wsoauth_user = 8 where wsoauth_user = 4;`  (old=4 -> new=8)

https://www.mediawiki.org/wiki/Extension:UserMerge/Hooks/MergeAccountFromTo needs to be implemented by WSOAuth here. -> DONE in my fork.

[^1]: WSOAuth feels a bit unfinished in some areas: the special page still talks about "_In the future, it will be possible to manage your remote accounts from this special page._", but that page hasn't changed in recent years . Then there's no information available to a wiki user to which external account a wiki account is currently linked to. I guess the plan was to include such information on the Special:WSOAuthConnectRemote page, along with an option to unlink an external account.

**Comment by mmd-osm** [comment]
@Firefishy : Since https://test.wiki.openstreetmap.org/ is still around, I would propose to give the changes in https://codeberg.org/mmd/chef (branch test/wikisso) a try. Some adjustments for the test wiki environment would be needed, as well as registering a new OAuth2 application.

This pulls in a WSOAuth fork in https://codeberg.org/mmd/WSOAuth as extension (branch sso). All changes can be conveniently reviewed in the [comparison with master](https://codeberg.org/mmd/WSOAuth/compare/master...osm).

I am testing all of this locally in Docker with Kitchen Converge, which is not ideal for getting more feedback from other people. The plan is to eventually get these changes merged upstream once testing has confirmed that all relevant scenarios are covered and working as expected.

We might want to pull in user email addresses at a later point in time[^1]. However, I can't test this scenario on osm.org due to lack of authorization to create a new OAuth2 app with email scope. For testing, users would need to enter their email address again on the Wiki and confirm it.

I would also suggest to deploy the [UserMerge](https://www.mediawiki.org/wiki/Extension:UserMerge) extension as @1ec5 has proposed.

No timelines, this is not urgent in any way.

Thanks!

[^1]: Relevant parts of the code that would have to adjusted: [link 1](https://codeberg.org/mmd/WSOAuth/src/branch/osm/src/AuthenticationProvider/OpenStreetMapAuth.php#L157) and [link 2](https://codeberg.org/mmd/WSOAuth/src/branch/osm/src/AuthenticationProvider/OpenStreetMapAuth.php#L106)

**Comment by mmd-osm** [comment]
I added some code to display the current "wiki user -> osm.org user" mapping on the Special page.

The red "Unassign" button deletes this mapping. Be sure your wiki user has an email address and password defined upfront, otherwise you'll lock yourself out. This isn't being checked right now.

The level of detail on this page is meant to help with testing/debugging. It's not the final version for sure.

<img width="1174" height="523" alt="Image" src="https://github.com/user-attachments/assets/5ffa162a-72ea-42b5-a558-23b4a815cb43" />

When creating new users with OSM, the same rules apply as before:

<img width="515" height="554" alt="Image" src="https://github.com/user-attachments/assets/3cdccd72-9047-4d05-a9de-5455a3c20964" />

**Comment by GioviPen** [comment]
https://oauth.net/2/

https://meta.wikimedia.org/wiki/Special:OAuth/authorize

https://www.mediawiki.org/wiki/Special:MyLanguage/Extension:OAuth

https://www.mediawiki.org/wiki/Special:MyLanguage/OAuth/For_Developers
